OperationBase::name() segfault when called immediately after Operation creation

m2es3h opened the pull request:
On construction of an OperationBase instance, `std::shared_ptr<Channel> OperationBase::chan` is nullptr until the Channel is constructed. Channel construction happens in a lambda function passed to `context->tcp_loop.dispatch([](){ chan = Channel::Build(...)})`.

In the time between OperationBase construction and deferred Channel construction, calling `OperationBase::name()` dereferences the empty shared_ptr to get `chan->name`.

This fix checks for nullptr first and returns a `std::string&` to a default string instead if needed.

mdavidsaver commented:
Confirmed.

mdavidsaver commented:
Not crashing is certainly an improvment!

Really though, `Operation::name()` should be able to return the channel name immediately.  Indirectly to the `Channel` could be seen as a premature optimization on my part.  And I was not even consistent (cf. `SubscriptionImpl::channelName`).  So I think the straightforward solution would be to store a channel name directly in OperationBase.

m2es3h commented:
> ... So I think the straightforward solution would be to store a channel name directly in OperationBase.

Makes sense, I've updated the PR to do this. I've also changed the return types of `name()` to copies instead of references, since using string references that outlive the OperationBase would also cause a segfault.
